### ISSUE TITLE
Realm private headers accessible from public headers

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/realm#165]
If I understand correctly, `realm/cuda/cuda_redop.h` and `realm/hip/hip_redop.h` are intended to be private headers; but there is a path to them from the public `realm.h`. As a result, any consumer of Realm ends up inheriting Realm's otherwise private CUDA/HIP dependency.

The practical consequence of this is that all of our Realm tests (and sometimes Legion ones) need to set their own CUDA/HIP include/link flags. A failure to do this looks like the following:

```
In file included from /gpfs/alpine/csc335/proj-shared/ci/572_1839/runtime/realm/redop.h:28,
                 from /gpfs/alpine/csc335/proj-shared/ci/572_1839/runtime/realm/runtime.h:22,
                 from /gpfs/alpine/csc335/proj-shared/ci/572_1839/runtime/realm/machine.inl:17,
                 from /gpfs/alpine/csc335/proj-shared/ci/572_1839/runtime/realm/machine.h:291,
                 from /gpfs/alpine/csc335/proj-shared/ci/572_1839/runtime/realm/instance.inl:27,
                 from /gpfs/alpine/csc335/proj-shared/ci/572_1839/runtime/realm/instance.h:355,
                 from /gpfs/alpine/csc335/proj-shared/ci/572_1839/runtime/realm/profiling.h:29,
                 from /gpfs/alpine/csc335/proj-shared/ci/572_1839/runtime/realm.h:43,
                 from /gpfs/alpine/csc335/proj-shared/ci/572_1839/test/realm/ctxswitch.cc:1:
/gpfs/alpine/csc335/proj-shared/ci/572_1839/runtime/realm/hip/hip_redop.h:22:10: fatal error: hip/hip_runtime.h: No such file or directory
   22 | #include <hip/hip_runtime.h>
      |          ^~~~~~~~~~~~~~~~~~~
```

Corresponding CI job: https://code.olcf.ornl.gov/ci/csc335/dev/legion/-/jobs/1839#L522

If we want users to be able to include `realm.h` without pulling in CUDA/HIP, we'll need to enforce that private headers are not reachable from public ones.

**Comment by muraj** [comment]
This should eventually be fixed with the work done on the C API and the C++ wrapper for the C API.  Once these are solidified, only these headers will be installed and exposed, so no more internal headers will be available.

In any case, the original issue seems more related to some confusion about what is public and what is not?  I think regardless of what we end up doing with the new C and C++ headers, in order to maintain compatibility, we will always have a redop.h, which will likely always have a cuda_redop.h and hip_redop.h which will also likely need the cuda_runtime.h / hip/hip.h header files for the GPU.

TL;DR - I think this is an invalid bug as the experience is as expected at this time.  @elliottslaughter can you confirm?

**Comment by elliottslaughter** [comment]
It's been so long that I don't remember how I worked around this. (Frontier CI is not currently working, but it did work for a long time after I filed this.)

However if I understand correctly the root cause is still there: *any* include of `realm.h` results (transitively) in `realm/redop.h` being included, which means that *any* Realm application has to include the relevant CUDA/HIP headers even if it does not itself use the GPU. E.g., the original report shows it being pulled in from `ctxswitch` which is not a GPU test as far as I recall.

Would you say that's working as intended? Because it still seems like an odd behavior to me.

**Comment by muraj** [comment]
> Would you say that's working as intended? Because it still seems like an odd behavior to me

I agree it's odd behavior... I'm trying to think of an analog here from other libraries and am failing.  As an example from cuda, cuda has opengl interop that requires opengl headers, but it's a separate header you include, it's not part of cuda.h, etc.  Anything that doesn't require those headers has its types forward declared and apis defined.

I think what we need to do is rework the redop objects' internal definition entirely as it is exposed in the header files.  For example, we should just unconditionally define all the function pointer definitions we need and not do this template metaprogramming magic to try to minimize the size of the redop structure for some apparent reason.

@eddy16112 marking this as part of the C API effort, as that will be needed to define the right interface, then when we build the c++ wrapper we'll have a pretty good idea of what we really need to do here.
